File: Ruby/examples/template_utils_example.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Template Engine Utilities Example
# Demonstrates various features of the template engine

require_relative '../template_utils/mod'

puts "=" * 60
puts "Template Engine Utilities - Examples"
puts "=" * 60

# Example 1: Basic variable interpolation
puts "\n1. Basic Variable Interpolation"
puts "-" * 40
template = "Hello, {{ name }}! Welcome to {{ place }}."
context = { name: "Alice", place: "Ruby World" }
result = AllToolkit::TemplateUtils.render(template, context)
puts "Template: #{template}"
puts "Context:  #{context}"
puts "Result:   #{result}"

# Example 2: Filters
puts "\n2. Filters"
puts "-" * 40
template = <<~TEMPLATE
Name: {{ name | upcase }}
Description: {{ desc | truncate: 20 }}
Price: ${{ price | round: 2 }}
Items: {{ items | join: ", " }}
Count: {{ items | size }}
TEMPLATE

context = {
  name: "product",
  desc: "This is a very long description that will be truncated",
  price: 19.999,
  items: ["red", "green", "blue"]
}
result = AllToolkit::TemplateUtils.render(template, context)
puts "Template:\n#{template}"
puts "Result:\n#{result}"

# Example 3: Conditionals
puts "\n3. Conditionals"
puts "-" * 40
template = <<~TEMPLATE
{% if user %}
Hello, {{ user }}!
{% else %}
Welcome, guest!
{% endif %}

{% if items %}
You have {{ items | size }} items.
{% else %}
Your cart is empty.
{% endif %}
TEMPLATE

context = { user: "Alice", items: ["a", "b", "c"] }
result = AllToolkit::TemplateUtils.render(template, context)
puts "Template:\n#{template}"
puts "Context: #{context}"
puts "Result:\n#{result}"

# Empty cart example
context = { user: nil, items: [] }
result = AllToolkit::TemplateUtils.render(template, context)
puts "\nContext: #{context}"
puts "Result:\n#{result}"

# Example 4: Loops
puts "\n4. Loops"
puts "-" * 40
template = <<~TEMPLATE
Products:
{% for product in products %}
- {{ product.name }}: ${{ product.price | round: 2 }}
{% endfor %}

Total: {{ products | size }} products
TEMPLATE

context = {
  products: [
    { name: "Apple", price: 1.5 },
    { name: "Banana", price: 0.75 },
    { name: "Cherry", price: 2.99 }
  ]
}
result = AllToolkit::TemplateUtils.render(template, context)
puts "Template:\n#{template}"
puts "Result:\n#{result}"

# Example 5: Comments
puts "\n5. Comments"
puts "-" * 40
template = <<~TEMPLATE
{# This is a comment and won't appear in output #}
Visible text
{# Comments can span
multiple lines #}
More visible text
TEMPLATE

result = AllToolkit::TemplateUtils.render(template, {})
puts "Template:\n#{template}"
puts "Result:\n#{result}"

# Example 6: Partials/Include
puts "\n6. Partials/Include"
puts "-" * 40
partials = {
  "header" => "=== {{ title | upcase }} ===",
  "footer" => "--- Generated at {{ time }} ---",
  "item_row" => "- {{ name }}: {{ status | default: 'pending' }}"
}

template = <<~TEMPLATE
{% include 'header' %}

Items:
{% for item in items %}
{% include 'item_row' %}
{% endfor %}

{% include 'footer' %}
TEMPLATE

context = {
  title: "report",
  time: "2024-01-15",
  items: [
    { name: "Task 1", status: "done" },
    { name: "Task 2", status: nil },
    { name: "Task 3", status: "in progress" }
  ]
}
result = AllToolkit::TemplateUtils.render(template, context, partials)
puts "Template:\n#{template}"
puts "Partials: #{partials.keys}"
puts "Result:\n#{result}"

# Example 7: HTML Escaping
puts "\n7. HTML Escaping"
puts "-" * 40
template = <<~TEMPLATE
Original: {{ html }}
Escaped:  {{ html | escape_html }}
Stripped: {{ html | strip_html }}
TEMPLATE

context = { html: "<script>alert('XSS')</script><p>Safe content</p>" }
result = AllToolkit::TemplateUtils.render(template, context)
puts "Template:\n#{template}"
puts "Result:\n#{result}"

# Example 8: Complex Example - Email Template
puts "\n8. Complex Example - Email Template"
puts "-" * 40
email_template = <<~TEMPLATE
Subject: {{ subject | titleize }}

Dear {{ customer.name | capitalize }},

Thank you for your order!

Order Details:
{% for item in order.items %}
- {{ item.name }} x{{ item.qty }} @ ${{ item.price | round: 2 }} = ${{ item.qty | times: item.price | round: 2 }}
{% endfor %}

{% if order.discount > 0 %}
Discount: -${{ order.discount | round: 2 }}
{% endif %}

Total: ${{ order.total | round: 2 }}

{% if order.total > 100 %}
Congratulations! You qualify for free shipping!
{% else %}
Shipping: ${{ order.shipping | round: 2 }}
{% endif %}

Best regards,
{{ company.name }}
{{ company.email }}
TEMPLATE

# Custom filter for multiplication
AllToolkit::TemplateUtils.register_filter(:times) do |value, multiplier|
  value.to_f * multiplier.to_f
end

context = {
  subject: "order confirmation",
  customer: { name: "john doe" },
  order: {
    items: [
      { name: "Widget", qty: 2, price: 29.99 },
      { name: "Gadget", qty: 1, price: 49.99 },
      { name: "Thingama", qty: 3, price: 9.99 }
    ],
    discount: 10.00,
    total: 139.94,
    shipping: 9.99
  },
  company: {
    name: "Acme Corp",
    email: "support@acme.com"
  }
}

result = AllToolkit::TemplateUtils.render(email_template, context)
puts "Template:\n#{email_template}"
puts "Result:\n#{result}"

# Example 9: Comparison Operators
puts "\n9. Comparison Operators"
puts "-" * 40
template = <<~TEMPLATE
{% if score >= 90 %}A{% elsif score >= 80 %}B{% elsif score >= 70 %}C{% else %}F{% endif %} Grade
Score: {{ score }}/100
{% if score == 100 %}Perfect!{% endif %}
{% if score != 0 %}Not zero{% endif %}
{% if score < 60 %}Failing{% else %}Passing{% endif %}
TEMPLATE

context = { score: 85 }
result = AllToolkit::TemplateUtils.render(template, context)
puts "Template:\n#{template}"
puts "Context: #{context}"
puts "Result:\n#{result}"

# Example 10: Using top-level convenience method
puts "\n10. Convenience Method"
puts "-" * 40
template = "Quick render: {{ msg | upcase }}"
result = render_template(template, { msg: "hello world" })
puts "Template: #{template}"
puts "Result:   #{result}"

puts "\n" + "=" * 60
puts "All examples completed!"
puts "=" * 60

File: TypeScript/examples/template_utils_example.ts
/**
 * TypeScript Template Engine Example
 * Demonstrates usage of the Template Engine Utilities
 */

import { TemplateEngine, renderTemplate } from '../template_utils/mod';

console.log('=== TypeScript Template Engine Examples ===\n');

// Example 1: Basic Variable Interpolation
console.log('1. Basic Variable Interpolation');
console.log('-------------------------------');
const greeting = renderTemplate('Hello, {{ name }}!', { name: 'World' });
console.log(`Template: "Hello, {{ name }}!"`);
console.log(`Context: { name: 'World' }`);
console.log(`Result: "${greeting}"\n`);

// Example 2: Filters
console.log('2. Using Filters');
console.log('----------------');
const upperName = renderTemplate('{{ name | upper }}', { name: 'alice' });
console.log(`Uppercase: "${upperName}"`);

const titleName = renderTemplate('{{ name | title }}', { name: 'hello world' });
console.log(`Title case: "${titleName}"`);

const trimmed = renderTemplate('{{ text | trim }}', { text: '  hello  ' });
console.log(`Trimmed: "${trimmed}"`);

const escaped = renderTemplate('{{ html | escape }}', { html: '<script>alert("xss")</script>' });
console.log(`Escaped HTML: "${escaped}"\n`);

// Example 3: Filter Chains
console.log('3. Filter Chains');
console.log('----------------');
const chained = renderTemplate('{{ text | trim | upper }}', { text: '  hello world  ' });
console.log(`Template: "{{ text | trim | upper }}"`);
console.log(`Result: "${chained}"\n`);

// Example 4: Conditionals
console.log('4. Conditionals');
console.log('---------------');
const conditionalTemplate = `{% if user %}
Hello, {{ user }}!
{% else %}
Welcome, guest!
{% endif %}`;

console.log('With user:');
const withUser = renderTemplate(conditionalTemplate, { user: 'Alice' });
console.log(withUser.trim());

console.log('\nWithout user:');
const withoutUser = renderTemplate(conditionalTemplate, {});
console.log(withoutUser.trim());
console.log();

// Example 5: Loops
console.log('5. Loops');
console.log('--------');
const loopTemplate = `Items:
{% for item in items %}
- {{ item }}
{% endfor %}`;

const loopResult = renderTemplate(loopTemplate, { items: ['Apple', 'Banana', 'Cherry'] });
console.log(loopResult);

// Example 6: Array Filters
console.log('6. Array Filters');
console.log('----------------');
const joined = renderTemplate('{{ items | join:", " }}', { items: ['red', 'green', 'blue'] });
console.log(`Joined: "${joined}"`);

const firstItem = renderTemplate('First: {{ items | first }}', { items: ['a', 'b', 'c'] });
console.log(`First item: "${firstItem}"`);

const lastItem = renderTemplate('Last: {{ items | last }}', { items: ['a', 'b', 'c'] });
console.log(`Last item: "${lastItem}"`);

const size = renderTemplate('Count: {{ items | size }}', { items: [1, 2, 3, 4, 5] });
console.log(`Size: "${size}"\n`);

// Example 7: Number Filters
console.log('7. Number Filters');
console.log('-----------------');
const rounded = renderTemplate('Price: ${{ price | round:2 }}', { price: 19.999 });
console.log(`Rounded: "${rounded}"`);

const absolute = renderTemplate('Absolute: {{ num | abs }}', { num: -42 });
console.log(`Absolute: "${absolute}"\n`);

// Example 8: Default Values
console.log('8. Default Values');
console.log('-----------------');
const withDefault = renderTemplate('Name: {{ name | default:Anonymous }}', {});
console.log(`With default: "${withDefault}"`);

const withValue = renderTemplate('Name: {{ name | default:Anonymous }}', { name: 'Alice' });
console.log(`With value: "${withValue}"\n`);

// Example 9: Partials (Includes)
console.log('9. Partials (Includes)');
console.log('----------------------');
const partials = {
    header: '<header><h1>{{ title }}</h1></header>',
    footer: '<footer>Copyright {{ year }}</footer>'
};

const pageTemplate = `{% include "header" %}
<main>Welcome to {{ site }}</main>
{% include "footer" %}`;

const pageResult = renderTemplate(pageTemplate, { title: 'My Site', site: 'example.com', year: 2024 }, partials);
console.log(pageResult);
console.log();

// Example 10: Custom Filters
console.log('10. Custom Filters');
console.log('------------------');
const engine = new TemplateEngine();
engine.registerFilter('double', (n: number) => n * 2);
engine.registerFilter('exclaim', (s: string) => s + '!!!');

const customResult = engine.render('{{ num | double }} and {{ text | exclaim }}', { num: 21, text: 'hello' });
console.log(`Custom filters result: "${customResult}"\n`);

// Example 11: Nested Objects
console.log('11. Nested Objects');
console.log('------------------');
const nestedTemplate = `User: {{ user.name }}
Email: {{ user.email }}
City: {{ user.address.city }}`;

const nestedContext = {
    user: {
        name: 'Alice',
        email: 'alice@example.com',
        address: {
            city: 'New York',
            country: 'USA'
        }
    }
};

const nestedResult = renderTemplate(nestedTemplate, nestedContext);
console.log(nestedResult);
console.log();

// Example 12: Comments
console.log('12. Comments');
console.log('------------');
const withComments = renderTemplate('Hello {# this is ignored #} World', {});
console.log(`Template with comments: "${withComments}"\n`);

// Example 13: Comparison Operators
console.log('13. Comparison Operators');
console.log('------------------------');
const eqResult = renderTemplate('{% if count == 5 %}Exactly five{% endif %}', { count: 5 });
console.log(`Equal: "${eqResult}"`);

const gtResult = renderTemplate('{% if count > 5 %}More than five{% endif %}', { count: 10 });
console.log(`Greater than: "${gtResult}"`);

const notResult = renderTemplate('{% if not user %}Guest{% endif %}', { user: '' });
console.log(`Not operator: "${notResult}"\n`);

// Example 14: URL Encoding
console.log('14. URL Encoding');
console.log('----------------');
const urlEncoded = renderTemplate('{{ text | url_encode }}', { text: 'hello world & more' });
console.log(`URL encoded: "${urlEncoded}"\n`);

// Example 15: Email Template Example
console.log('15. Email Template');
console.log('------------------');
const emailTemplate = `Subject: {{ subject | upper }}

Dear {{ user.name | title }},

Thank you for your order of {{ items | size }} items:
{% for item in items %}
- {{ item.name }} (x{{ item.quantity }}) - ${{ item.price }}
{% endfor %}

Total: ${{ total | round:2 }}

{% if discount > 0 %}
You saved ${{ discount | round:2 }}!
{% endif %}

Best regards,
The Team`;

const emailContext = {
    subject: 'order confirmation',
    user: { name: 'john doe' },
    items: [
        { name: 'Widget', quantity: 2, price: 29.99 },
        { name: 'Gadget', quantity: 1, price: 49.99 }
    ],
    total: 109.97,
    discount: 10.00
};

const emailResult = renderTemplate(emailTemplate, emailContext);
console.log(emailResult);

console.log('\n=== All Examples Complete ===');
